pep-9999.rst

```diff
@@ -0,0 +1,669 @@
+freezing a class needs to freeze its superclasses as well.
+
+
+Subclassing Immutable Classes
+-----------------------------
+
+CPython classes hold references to their subclasses.
+If immutability it taken literally, it would not be
+permitted to create a subclass of an immutable type.
+Because this reference is "accidental" and does not
+get exposed to the programmer in any dangerous way,
+we permit frozen classes to be subclassed (by mutable
+classes). C.f. `Sharing Immutable Data Across Subinterpreters`_.
+
+
+Freezing Function Objects
+-------------------------
+
+Freezing function objects and lambdas is suprisingly involved
+because all function objects have a pointer to ``globals``.
+Function objects can be thought of as regular objects whose
+fields are its local variables -- some of which may be captured
+from enclosing scopes.
+
+Consider the following scenario:
+
+.. code-block:: python
+
+	x = 0
+	def foo():
+		return x
+
+	freeze(foo)
+	... # some code, e.g. pass foo to another thread
+	x = 1
+	foo()
+
+In the code above, the ``foo`` function object captures the ``x``
+variable from its enclosing scope. While ``x`` happens to point to
+an immutable object, the variable itself (the frame of the function object)
+is mutable. Unless something is done to prevent it (see below!), passing
+``foo`` to another thread will make the assignment ``x = 1`` a potential
+data race.
+
+We consider freezing of a function to freeze that function's
+meaning at that point in time. In the code above, that means that
+``foo`` gets its own copy of ``x`` which will have value of the
+original at the time of freezing, in this case 0.
+
+Thus, the assignment ``x = 1`` is still permitted as it will not affect
+``foo``, and it may therefore not contribute to a data race. Furthermore,
+the result of calling ``foo()`` will be 0 -- not 1!
+
+This can be implemented by having ``x`` in ``foo`` point to a
+fresh cell and then freezing the cell (and similar for global capture).
+
+We believe that this design is a sweet-spot that is intuitive and
+permissive.
+
+Now consider freezing the following function:
+
+.. code-block:: python
+
+	x = 0
+	def foo(a = False):
+		if a:
+			a = a + 1
+			return a
+		else:
+			x = x + 1
+			return x
+
+	freeze(foo)
+	foo(41)  # OK, returns 42
+	foo()    # Throws NotWriteableError
+
+This example illustrates two things. The first call to ``foo(41)``
+shows that local variables on the frame of a frozen function are
+mutable. The second call shows that captured variables are not.
+Note that the default value of ``a`` will be frozen when ``foo``
+is frozen. Thus, the problem of side-effects on default values
+on parameters is avoided.
+
+Frozen function objects that access globals, e.g. through a call
+to ``globals()`` will throw an exception when called.
+
+
+Implementation Details
+======================
+
+1. Introduce an ``is_immutable`` property on objects that tracks whether or
```

Comment:
I assume "property" here is meant in the abstract sense.  To avoid confusion with Python object properties, maybe say "concept"?